Sources/audio-devices/AudioDevices.swift

```diff
@@ -1,126 +1,249 @@
+      
+      return Double(deviceVolume)
+    } catch {
+      return nil
+    }
+  }
 
-      do {
-        try CoreAudioData.get(
-          id: deviceId,
-          selector: kAudioDevicePropertyVolumeScalar,
-          scope: kAudioDevicePropertyScopeOutput,
-          value: &deviceVolume
-        )
+  var outputVolume: Double? {
+    do {
+      var deviceVolume: Float32 = 0.0
+      try CoreAudioData.get(
+        id: id,
+        selector: kAudioHardwareServiceDeviceProperty_VirtualMasterVolume,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &deviceVolume
+      )
+      
+      return Double(deviceVolume)
+    } catch {
+      return nil
+    }
+  }
+  
+  var volume: Double? {
+    // TODO: need to think how to handle when hasInput && hasOutput
+    if hasInput {
+      return self.inputVolume
+    }
+    if hasOutput {
+      return self.outputVolume
+    }
+    
+    return nil
+  }
 
-        volume = Double(deviceVolume)
-      } catch {
-        volume = nil
+  func setVolume(to newVolume: Double, for type: AudioDeviceType) throws {
+    if (type.hasInput) {
+      guard (0...1).contains(newVolume) else {
+        throw AudioDevice.Error.invalidVolumeValue
       }
-    } else {
-      volume = nil
+      
+      var value = Float32(newVolume)
+      try CoreAudioData.set(
+        id: id,
+        selector: kAudioDevicePropertyVolumeScalar,
+        scope: kAudioDevicePropertyScopeInput,
+        value: &value
+      )
+    }
+    if (type.hasOutput) {
+      guard (0...1).contains(newVolume) else {
+        throw AudioDevice.Error.invalidVolumeValue
+      }
+      
+      var value = Float32(newVolume)
+      try CoreAudioData.set(
+        id: id,
+        selector: kAudioHardwareServiceDeviceProperty_VirtualMasterVolume,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &value
+      )
     }
   }
 
-  mutating func setVolume(_ newVolume: Double) throws {
-    guard volume != nil else {
-      throw Error.volumeNotSupported
+  func setVolume(to newVolume: Double) throws {
+    // TODO: need to think how to handle when hasInput && hasOutput
+    if hasInput {
+      return try setVolume(to: newVolume, for: .input)
+    }
+    if hasOutput {
+      return try setVolume(to: newVolume, for: .output)
+    }
+  }
+  
+  var isInputMuted: Bool? {
+    do {
+      var muteValue: UInt32 = 0
+      try CoreAudioData.get(
+        id: id,
+        selector: kAudioDevicePropertyMute,
+        scope: kAudioDevicePropertyScopeInput,
+        value: &muteValue
+      )
+      
+      return muteValue != 0
+    } catch {
+      return nil
+    }
+  }
+  
+  var isOutputMuted: Bool? {
+    do {
+      var muteValue: UInt32 = 0
+      try CoreAudioData.get(
+        id: id,
+        selector: kAudioDevicePropertyMute,
+        scope: kAudioDevicePropertyScopeOutput,
+        value: &muteValue
+      )
+      
+      return muteValue != 0
+    } catch {
+      return nil
+    }
+  }
+  
+  var isMuted: Bool? {
```

Comment:
Should we just default to `false` here instead of making it optional? Optional booleans is generally an anti-pattern.